core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -305,20 +312,39 @@ public void registerFunctionScoreParser(ScoreFunctionParser<? extends ScoreFunct
      *
      * @param reader the reader registered for this query's builder. Typically a reference to a constructor that takes a
      *        {@link org.elasticsearch.common.io.stream.StreamInput}
-     * @param parser the parser the reads the query builder from xcontent
-     * @param names all names by which this query might be parsed. The first name is special as it is the name by under which the reader is
-     *        registered. So it is the name that the query should use as its {@link NamedWriteable#getWriteableName()}.
+     * @param queryParser the parser the reads the query builder from xcontent
+     * @param queryName holds the names by which this query might be parsed. The {@link ParseField#getPreferredName()} is special as it
+     *        is the name by under which the reader is registered. So it is the name that the query should use as its
+     *        {@link NamedWriteable#getWriteableName()} too.
      */
-    public <QB extends QueryBuilder<QB>> void registerQuery(Writeable.Reader<QB> reader, QueryParser<QB> parser, String... names) {
-        queries.add(new QueryRegistration<QB>(names, reader, parser));
+    public synchronized <QB extends QueryBuilder<QB>> void registerQuery(Writeable.Reader<QB> reader, QueryParser<QB> queryParser,
+                                                                         ParseField queryName) {
+        innerRegisterQueryParser(queryParser, queryName);
+        namedWriteableRegistry.register(QueryBuilder.class, queryName.getPreferredName(), reader);
     }
 
     /**
      * Register a query.
      * TODO remove this in favor of registerQuery
      */
-    public void registerQueryParser(Supplier<QueryParser<?>> parser) {
-        queryParsers.add(parser);
+    public void registerQueryParser(QueryParser<?> queryParser, ParseField queryName) {
+        innerRegisterQueryParser(queryParser, queryName);
+        namedWriteableRegistry.registerPrototype(QueryBuilder.class, queryParser.getBuilderPrototype());
+    }
+
+    private <QB extends QueryBuilder<QB>> void innerRegisterQueryParser(QueryParser<QB> parser, ParseField queryName) {
+        Tuple<ParseField, QueryParser<?>> parseFieldQueryParserTuple = new Tuple<>(queryName, parser);
```

Comment:
I changed my mind, I think the register methods should be only here in SearchModule. The nice thing of IndicesQueriesRegistry is that it's immutable, and classes that don't depend on it do so only because they need to read from it, retrieve registered parsers. I wouldn't want to expose a register method there and make the class mutable. You see what I mean?